### PR TITLE
Lazy TextField

### PR DESCRIPTION
**Pull request by iitzco-asapp**
Because of how serialization works, the `TextField` will be initialized each time it is de-serialized.

Because of this behavior, we don't want the embedding model to be loaded in the init, mainly because the embedding model is not used when `setup` is not needed (this is the case for inference models)

This PR initializes the embedding model in the `setup` method rather than in the init, keeping object initialization super lightweight.

### CODE REVIEW

File: flambe/field/text.py [hunk 1]
```diff
@@ -65,6 +65,46 @@ def get_embeddings(
...
+        Whether to drop tokens that don't have embeddings
+        associated. Defaults to True.
+        Important: this flag will only work when using embeddings.
+
+    """
+    embeddings: str
+    embeddings_format: str = 'gensim'
+    embeddings_binary: bool = False
+    setup_all_embeddings: bool = False
+    unk_init_all: bool = False
```
Comment (jeremyasapp) [inline flambe/field/text.py]: this is not necessary embedding related. Neither is drop_unknown

File: flambe/field/text.py [hunk 2]
```diff
@@ -90,10 +130,10 @@ def __init__(self,  # nosec
                  unk_token: str = '<unk>',
                  sos_token: Optional[str] = None,
                  eos_token: Optional[str] = None,
+                 embeddings_info: Optional[EmbeddingsInformation] = None,
```
Comment (jeremyasapp) [inline flambe/field/text.py]: Why did you add this? Seems unnecessary, and adds an extra level of nesting

File: flambe/field/text.py (hunk 1 above)
Comment (iitzco-asapp) [reply flambe/field/text.py]: iirrc and according to the documentation, this parameters are only used in case embeddings are provided. Do you have an example where they are not used with embeddings?

Comment (iitzco-asapp) [reply flambe/field/text.py]: `unk_init_all` is for for the vocab that is not in the embeddings.

Comment (jeremyasapp) [reply flambe/field/text.py]: well for drop_unkown you can drop things that are not in the vocab, and you may have created the vocab from the data not the embeddings file

File: flambe/field/text.py [hunk 3]
```diff
@@ -65,6 +65,46 @@ def get_embeddings(
...
+    drop_unknown: bool
+        Whether to drop tokens that don't have embeddings
+        associated. Defaults to True.
+        Important: this flag will only work when using embeddings.
+
+    """
+    embeddings: str
+    embeddings_format: str = 'gensim'
+    embeddings_binary: bool = False
+    setup_all_embeddings: bool = False
```
Comment (jeremyasapp) [inline flambe/field/text.py]: I don't like this parameter name, it's not clear that it has to do with using the embeddings as source of the vocabulary

File: flambe/field/text.py (hunk 1 above)
Comment (iitzco-asapp) [reply flambe/field/text.py]: and `drop_unknown` just works now when embeddings are passed. Maybe we can generalize this one to work also for non embedding TextFields

File: flambe/field/text.py (hunk 2 above)
Comment (iitzco-asapp) [reply flambe/field/text.py]: By having all related parameters in the same data structure, the code becomes:

* more clear and readable, following a composition pattern allowing the creation of factories 
* easier to test
* easier to extend the class as your __init__ doesn't have tons of parameters anymore
* easier to implement custom logic in this new class rather than in the TextField iteself. For example, avoiding having in the TextField checks like "user passed the `embedding` name but `embedding_format` is None" which can be delegated to the appropriate class.

Users won't have another level of nesting in they configs as they will use the factory

File: flambe/field/text.py (hunk 1 above)
Comment (iitzco-asapp) [reply flambe/field/text.py]: > well for drop_unkown you can drop things that are not in the vocab, and you may have created the vocab from the data not the embeddings file

Yes lets generalize this concept

File: flambe/field/text.py (hunk 3 above)
Comment (iitzco-asapp) [reply flambe/field/text.py]: `embeddings_as_vocab_source`? Better ideas?

Comment (brianasapp) [reply flambe/field/text.py]: `use_all_from_vectored_keys`?

File: flambe/field/text.py [hunk 4]
```diff
@@ -258,7 +325,7 @@ def _build_embeddings(self, model: KeyedVectors) -> Tuple[odict, torch.Tensor]:
 
         tokens: Iterable[str] = self.vocab.keys()
 
-        if self.setup_all_embeddings:
+        if setup_all_embeddings:
```
Comment (axel-asapp) [inline flambe/field/text.py]: I would rename to avoid the outer scope naming collapse. It would become clearer it is definitely a local variable

File: flambe/field/text.py [hunk 5]
```diff
@@ -272,7 +339,7 @@ def _build_embeddings(self, model: KeyedVectors) -> Tuple[odict, torch.Tensor]:
                 else:
                     self.unk_numericals.add(self.vocab[token])
 
-                    if self.unk_init_all:
+                    if unk_init_all:
```
Comment (axel-asapp) [inline flambe/field/text.py]: I would rename to avoid the outer scope naming collapse. It would become clearer it is definitely a local variable

File: flambe/field/text.py (hunk 2 above)
Comment (axel-asapp) [reply flambe/field/text.py]: I think is a great idea to take subclassing and logic extraction into appropiate components 👏

File: flambe/field/text.py (hunk 3 above)
Comment (iitzco-asapp) [reply flambe/field/text.py]: Lets do `build_vocab_from_embeddings`.

Already discussed with the team, this is a harmless backwards incompatible change.

Comment (iitzco-asapp) [reply flambe/field/text.py]: Actually, no backwards incompatibility as we keep `setup_all_embeddings` in the init
